### PR TITLE
Adding a new aliases defining logic for `SELECT` and `CREATE VIEW`

### LINKED ISSUE #70139: Support bracket aliases for VIEWs and SELECTs
Seen in TPC-H Q15

```sql
-- doesn't work
CREATE VIEW revenue0 (supplier_no, total_revenue) AS
    SELECT
        l_suppkey,
        sum(l_extendedprice * (1 - l_discount))
    FROM
        lineitem
    WHERE
        l_shipdate >= DATE '1996-01-01'
        AND l_shipdate < DATE '1996-01-01' + INTERVAL '3' MONTH
    GROUP BY
        l_suppkey;

-- works
CREATE VIEW revenue0 AS
    SELECT
        l_suppkey AS supplier_no,
        sum(l_extendedprice * (1 - l_discount)) AS total_revenue
    FROM
        lineitem
    WHERE
        l_shipdate >= DATE '1996-01-01'
        AND l_shipdate < DATE '1996-01-01' + INTERVAL '3' MONTH
    GROUP BY
        l_suppkey;
```

Smaller example

```sql
-- doesn't work:
CREATE VIEW my_view (a, b) AS 
SELECT 1, 2

-- needs to be manually rewritten to:
CREATE VIEW my_view AS 
SELECT 1 as a, 2 as b
```

Non-TPC-H example:

```sql
-- doesn't work:
SELECT a, b FROM
(
SELECT number, number*2
FROM numbers(10)
) as x (a, b)

-- manually rewritten:
SELECT a, b FROM
(
SELECT number as a, number*2 as b
FROM numbers(10)
) as x
```

### PR DESCRIPTION
**Pull request by yariks5s**
Resolves https://github.com/ClickHouse/ClickHouse/issues/70139

ClickHouse now supports aliases

```sql
SELECT a, b FROM
(
  SELECT number, number*2
  FROM numbers(10)
) AS x (a, b); --    <-- aliases
```

``` SQL
CREATE VIEW my_view (a, b) AS --     <-- aliases
SELECT 1, 2;
```

``` SQL
WITH t (a, b) AS (  --    <-- aliases
     SELECT 1, 2
 )
 SELECT a, b
 FROM t;
```

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
`SELECT` and `VIEW` statements now support aliases, e.g. `SELECT b FROM (SELECT number, number*2 FROM numbers(2)) AS x (a, b);`. This enables TPC-H query 15 to run without modifications.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: tests/queries/0_stateless/03280_aliases_override_logic.sql [hunk 1]
```diff
@@ -0,0 +1,27 @@
+SELECT b FROM
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03280_aliases_override_logic.sql:1]: Test name: 03280_aliases_for_selects_and_views.sql

File: tests/queries/0_stateless/03280_aliases_override_logic.sql [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+(
+    SELECT number, number*2
+    FROM numbers(2)
+) as x (a, b);
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03280_aliases_override_logic.sql:5]: Cosmetic: `as` --> `AS`

File: tests/queries/0_stateless/03280_aliases_override_logic.sql [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+SELECT a FROM
+(
+    SELECT number, number*2
+    FROM numbers(2)
+) as x (a, b);
+
+SELECT a FROM
+(
+SELECT number, number*2
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03280_aliases_override_logic.sql:15]: indentation missing

File: tests/queries/0_stateless/03280_aliases_override_logic.sql [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+FROM numbers(10)
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03280_aliases_override_logic.sql:16]: 10 --> 2 (for consistency with the previous two tests)

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 2f6936bf2afeccce5ef3bbcaf8da5f4db167b5e2 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72480/2f6936bf2afeccce5ef3bbcaf8da5f4db167b5e2/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/72480/2f6936bf2afeccce5ef3bbcaf8da5f4db167b5e2/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: tests/queries/0_stateless/03280_aliases_override_logic.sql [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+) as x (a); -- { serverError BAD_ARGUMENTS }
+
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03280_aliases_override_logic.sql:18]: Could you add a test for the case that we select a non-existing column?

```sql
SELECT c FROM
(
    SELECT number, number*2
    FROM numbers(2)
) as x (a, b);
```

File: tests/queries/0_stateless/03280_aliases_override_logic.sql [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+CREATE VIEW IF NOT EXISTS test_view_03280 (a,b) AS SELECT 1, 2;
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03280_aliases_override_logic.sql:19]: So if `test_view_03280` somehow exists already and it is defined differently, we'll run into trouble.

Better do `DROP VIEW IF EXISTS` followed by `CREATE VIEW` (without `IF EXISTS`)

File: tests/queries/0_stateless/03280_aliases_override_logic.sql [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+SELECT a FROM test_view_03280;
+
+SELECT b FROM test_view_03280;
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03280_aliases_override_logic.sql:23]: + `SELECT c FROM test_view_03280` (which is expected to be an error)

File: src/Parsers/ParserSelectQuery.cpp [hunk 8]
```diff
@@ -500,6 +511,7 @@ bool ParserSelectQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
     select_query->setExpression(ASTSelectQuery::Expression::WITH, std::move(with_expression_list));
     select_query->setExpression(ASTSelectQuery::Expression::SELECT, std::move(select_expression_list));
     select_query->setExpression(ASTSelectQuery::Expression::TABLES, std::move(tables));
+    select_query->setExpression(ASTSelectQuery::Expression::ALIASES_OVERRIDE, std::move(expression_list_after_from));
```
Comment (rschu1ze) [inline src/Parsers/ParserSelectQuery.cpp:514]: "ALIAS_OVERRIDE" sounds a bit weird to me. Can we just name this `ALIAS`?

File: src/Parsers/ParserSelectQuery.cpp [hunk 9]
```diff
@@ -72,6 +72,7 @@ bool ParserSelectQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
     ParserNotEmptyExpressionList exp_list(false);
     ParserNotEmptyExpressionList exp_list_for_with_clause(false);
     ParserNotEmptyExpressionList exp_list_for_select_clause(/*allow_alias_without_as_keyword*/ true, /*allow_trailing_commas*/ true);
+    ParserAliasesExpressionList exp_list_after_from;
```
Comment (rschu1ze) [inline src/Parsers/ParserSelectQuery.cpp:75]: Please also update
- https://clickhouse.com/docs/en/sql-reference/statements/select
- https://clickhouse.com/docs/en/sql-reference/statements/create/view

File: src/Parsers/ASTSelectQuery.h [hunk 10]
```diff
@@ -20,6 +20,7 @@ class ASTSelectQuery : public IAST
         WITH,
         SELECT,
         TABLES,
+        ALIASES_OVERRIDE,
```
Comment (rschu1ze) [inline src/Parsers/ASTSelectQuery.h:23]: Pls see my below comment about "override"

File: tests/queries/0_stateless/03280_aliases_override_logic.sql [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+DROP VIEW IF EXISTS test_view_03280;
+
+CREATE VIEW test_view_1_03280 (a) AS SELECT 1, 2; -- { serverError BAD_ARGUMENTS }
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03280_aliases_override_logic.sql:27]: Consider adding some degenerate cases like

`CREATE VIEW test_view_1_03280 (a, b] AS SELECT 1, 2;`

and

`CREATE VIEW test_view_1_03280 ((a, b)) AS SELECT 1, 2;`

File: docs/en/sql-reference/statements/select/index.md [hunk 12]
```diff
@@ -17,6 +17,7 @@ SELECT [DISTINCT [ON (column1, column2, ...)]] expr_list
 [SAMPLE sample_coeff]
 [ARRAY JOIN ...]
 [GLOBAL] [ANY|ALL|ASOF] [INNER|LEFT|RIGHT|FULL|CROSS] [OUTER|SEMI|ANTI] JOIN (subquery)|table (ON <expr_list>)|(USING <column_list>)
+[(column1_name [, column2_name ...])]
```
Comment (rschu1ze) [inline docs/en/sql-reference/statements/select/index.md:20]: Same comment as in view.md

**Comment by yariks5s** [comment]
For some reason, tests that expect `SYNTAX_ERROR` don't work when they are at the bottom of the file.

File: src/Parsers/ASTCreateQuery.h [hunk 13]
```diff
@@ -4,6 +4,7 @@
 #include <Parsers/ASTQueryWithOnCluster.h>
 #include <Parsers/ASTDictionary.h>
 #include <Parsers/ASTDictionaryAttributeDeclaration.h>
+#include <Parsers/ASTIdentifier.h>
```
Comment (rschu1ze) [inline src/Parsers/ASTCreateQuery.h:7]: Why is the include needed?

File: src/Parsers/ASTCreateQuery.h [hunk 14]
```diff
@@ -105,6 +106,7 @@ class ASTCreateQuery : public ASTQueryWithTableAndOutput, public ASTQueryWithOnC
     bool has_uuid{false}; // CREATE TABLE x UUID '...'
 
     ASTColumns * columns_list = nullptr;
+    ASTExpressionList * aliases_list = nullptr; // Explicit aliases for syntax like CREATE VIEW my_view (a, b) AS SELECT 1, 2
```
Comment (rschu1ze) [inline src/Parsers/ASTCreateQuery.h:109]: Lets make the comment a little bit easier to understand:
```cpp
[...] /// Aliases such as "(a, b)" in "CREATE VIEW my_view (a, b) AS SELECT 1, 2"

File: src/Parsers/ASTCreateQuery.cpp [hunk 15]
```diff
@@ -240,6 +240,8 @@ ASTPtr ASTCreateQuery::clone() const
 
     if (columns_list)
         res->set(res->columns_list, columns_list->clone());
+    if (aliases_list)
```
Comment (rschu1ze) [inline src/Parsers/ASTCreateQuery.cpp:243]: Don't we need to change something in `ASTCreateQuery::formatImpl` and `ASTCreateQuery::formatQueryImpl` to print the aliases if the formatted query?

File: src/Parsers/ASTSelectQuery.h [hunk 16]
```diff
@@ -20,6 +20,7 @@ class ASTSelectQuery : public IAST
         WITH,
         SELECT,
         TABLES,
+        ALIAS,
```
Comment (rschu1ze) [inline src/Parsers/ASTSelectQuery.h:23]: `ALIASES`?

File: src/Parsers/ASTSelectQuery.h [hunk 17]
```diff
@@ -96,6 +99,7 @@ class ASTSelectQuery : public IAST
 
     ASTPtr & refSelect()    { return getExpression(Expression::SELECT); }
     ASTPtr & refTables()    { return getExpression(Expression::TABLES); }
+    ASTPtr & refAliasesOverride()    { return getExpression(Expression::ALIAS); }
```
Comment (rschu1ze) [inline src/Parsers/ASTSelectQuery.h:102]: Could we remove all mentions of "Override"?

File: src/Parsers/ASTSelectQuery.h (hunk 16 above)
Comment (rschu1ze) [inline src/Parsers/ASTSelectQuery.h:23]: See my comment about `formatImpl` in ASTCreateQuery.cpp, it applies here too.

File: src/Parsers/ParserCreateQuery.cpp [hunk 18]
```diff
@@ -1560,15 +1562,33 @@ bool ParserCreateViewQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expec
         if (!table_name_p.parse(pos, to_table, expected))
             return false;
     }
+    bool aliases_override = false;
```
Comment (rschu1ze) [inline src/Parsers/ParserCreateQuery.cpp:1565]: `has_aliases`?

File: src/Parsers/ParserCreateQuery.cpp [hunk 19]
```diff
@@ -1560,15 +1562,33 @@ bool ParserCreateViewQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expec
...
 
     /// Optional - a list of columns can be specified. It must fully comply with SELECT.
-    if (s_lparen.ignore(pos, expected))
+    if (s_lparen.ignore(pos, expected)) // Parsing cases like CREATE VIEW (a Int64, b Int64) (a, b) SELECT ...
```
Comment (rschu1ze) [inline src/Parsers/ParserCreateQuery.cpp:1568]: So this is implemented in `ParserCreateViewQuery::parseImpl` which, according to l. 553 in src/Parsers/ParserCreateQuery.h, runs not only for `CREATE VIEW` but also for `CREATE TABLE` etc. Is this what we really want or should we restrict the code to views?

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 20]
```diff
@@ -934,6 +934,45 @@ InterpreterCreateQuery::TableProperties InterpreterCreateQuery::getTableProperti
...
+        {
+            auto & aliases_children = create.aliases_list->children;
+            const auto * select_with_union_query = create.select->as<ASTSelectWithUnionQuery>();
+
+            if (!select_with_union_query)
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Expected ASTSelectWithUnionQuery");
+
+            const auto & selects = select_with_union_query->list_of_selects->children;
+
+            for (const auto & select_query_ptr : selects)
```
Comment (rschu1ze) [inline src/Interpreters/InterpreterCreateQuery.cpp:947]: `const auto & select : `

File: src/Parsers/ParserCreateQuery.cpp (hunk 19 above)
Comment (yariks5s) [reply src/Parsers/ParserCreateQuery.cpp:1568]: it is the `ParserCreateViewQuery` class, on l. 622, which is responsible for creating views only

**Comment by alexey-milovidov** [comment]
Add a test with CTE: `WITH t(a, b) AS (SELECT ...)`

**Comment by yariks5s** [comment]
> Add a test with CTE: `WITH t(a, b) AS (SELECT ...)`

You mean that something like this:

``` SQL
WITH t (a, b) AS (
    SELECT 1, 2
)
SELECT a, b
FROM t
```

should work?

**Comment by alexey-milovidov** [comment]
Yes, this will be good!

**Review by rschu1ze — APPROVED** [review]
Please add a test case for @alexey-milovidov's example. Besides that, everything lgtm, thanks!

**Comment by yariks5s** [comment]
> Please add a test case for @alexey-milovidov's example. Besides that, everything lgtm, thanks!

Yes, I will add that use case, but it needs to apply a new logic to parsers, so I will do it it next commit

**Review by novikd — CHANGES_REQUESTED** [review]

**Comment by clickgapai** [comment]
Hi @yariks5s @alexey-milovidov @rschu1ze — while reviewing this PR I found the following:

- **Bug (P1):** `buildSelectIntersectExceptQuery` missing `aliases` parameter causes CTE column aliases to fail for INTERSECT/EXCEPT queries — https://github.com/ClickHouse/ClickHouse/issues/104770

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**